### PR TITLE
Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by RileyManda**
We have completed the task of Integration specs for views and fix n+1 problems to the Blog Application according to the stipulated [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md): 🌱 

# Implementations:  

- [x]  Used Capybara to test the users, comments and posts views

- [x]  Analyzed and Fixed n+1 problems

>  ### Analyzing N+1 : Users: 🔽 

Screen Shot 2023-10-09 at 1 10 16 PM

Screen Shot 2023-10-09 at 1 11 57 PM

>  ### Analyzing N+1: Posts: 🔽

Screen Shot 2023-10-09 at 1 20 07 PM

> ### Analyzing N+1 : Comments:

#### N+1 issue in comments **Identified**: Before: 🔽 

Screen Shot 2023-10-09 at 1 25 38 PM
Screen Shot 2023-10-09 at 1 24 10 PM

#### N+1 issue **Fix**: After: 🔽 

Screen Shot 2023-10-09 at 1 34 46 PM

Screen Shot 2023-10-09 at 1 38 16 PM

- [x] ### Implemented the following View Specs

> User index page:

-     I can see the username of all other users.
-     I can see the profile picture for each user.
-     I can see the number of posts each user has written.
-     When I click on a user, I am redirected to that user's show page.

> User show page:

-     I can see the user's profile picture.
-     I can see the user's username.
-     I can see the number of posts the user has written.
-     I can see the user's bio.
-     I can see the user's first 3 posts.
-     I can see a button that lets me view all of a user's posts.
-     When I click a user's post, it redirects me to that post's show page.
-     When I click to see all posts, it redirects me to the user's post's index page.

> User post index page:

-     I can see the user's profile picture.
-     I can see the user's username.
-     I can see the number of posts the user has written.
-     I can see a post's title.
-     I can see some of the post's body.
-     I can see the first comments on a post.
-     I can see how many comments a post has.
-     I can see how many likes a post has.
-     I can see a section for pagination if there are more posts than fit on the view.
-     When I click on a post, it redirects me to that post's show page.

> Post show page:

-     I can see the post's title.
-     I can see who wrote the post.
-     I can see how many comments it has.
-     I can see how many likes it has.
-     I can see the post body.
-     I can see the username of each commentor.
-     I can see the comment each commentor left.

 In addition We have ensured the following general requirements:
 
 [Linters](https://github.com/RileyManda/blog-app/actions/workflows/linters.yml)

> Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
> Made sure that I have used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
> Made sure that I have documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> Followed list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

Please kindly review this implementation.

Thank you 🙏🏻

### CODE REVIEW

File: spec/features/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  scenario 'expect #index to display view title' do
+    expect(page).to have_content('Users')
+  end
+
+  scenario 'displays users information' do
+    users_data.each do |user_data|
+      expect(page).to have_content(user_data[:name])
+      expect(page).to have_content(user_data[:posts_counter])
+      expect(page).to have_css("img[src*='profile-image']")
```
Comment (rloterh) [inline spec/features/users_index_spec.rb]: - [ ] A test to assert the scenario **_When I click on a user, I am redirected to that user's show page._** for _user index view_ integration spec is missing! Please ensure to improve upon this aspect.

File: spec/features/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  def visit_user_show_page(user)
+    visit user_path(user)
+  end
+
+  def expect_posts_to_be_displayed(posts)
+    posts.each do |post_data|
+      expect(page).to have_content(post_data[:title])
+      expect(page).to have_content(post_data[:text])
```
Comment (rloterh) [inline spec/features/users_show_spec.rb]: - [ ] A test to assert the scenario **_When I click a user's post, it redirects me to that post's show page_** for _user show view_ integration spec is missing! Please ensure to improve upon this aspect.

File: spec/features/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    users_data.each do |user_data|
+      user = User.create(user_data)
+      posts_data.each { |post_data| user.posts.create(post_data) }
+    end
+  end
+
+  def visit_user_show_page(user)
+    visit user_path(user)
+  end
+end
```
Comment (rloterh) [inline spec/features/posts_index_spec.rb]: - [ ] Let's ensure to add the following test scenarios for _post index_ views:
  * I can see the user's profile picture
  * I can see the user's username
  * I can see the number of posts the user has written
  * I can see a post's title
  * I can see some of the post's body
  * I can see the first comments on a post
  * I can see how many comments a post has
  * I can see how many likes a post has
  * I can see a section for pagination if there are more posts than fit on the view

File: spec/features/posts_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  def expect_posts_to_be_displayed(posts)
+    posts.each do |post_data|
+      expect(page).to have_content(post_data[:title])
+      expect(page).to have_content(post_data[:text])
+      expect(page).to have_content("Comments: #{post_data[:comments_counter]}")
+      expect(page).to have_content("Likes: #{post_data[:likes_counter]}")
+    end
+  end
```
Comment (rloterh) [inline spec/features/posts_show_spec.rb]: - [ ] Let's ensure to add the following test scenarios for _post show_ views:
  * I can see the post's title
  * I can see how many comments it has
  * I can see how many likes it has
  * I can see the post body
  * I can see the username of each commentor
  * I can see the comment each commentor left

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Riley, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project, but you are almost there! 
🏁 🚦  ⛪ 🚜  🏃‍♀️  🌴  🏫 🌳 🏘️ 

## Highlights 👍  
- N + 1 problem addressed!
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by RileyManda** [comment]
### Iterations
I have completed the changes requested by the reviewer:

#### Implementations
 
- [x] Added spec to user_index: test to assert the scenario When I click on a user, I am redirected to that user's show page.

**Comment by RileyManda** [comment]
The following iterations have been implemented as requested by the reviewer

#### Iterations


## post/index

- can see the user's profile picture
- can see the user's username
- can see the number of posts the user has written
- can see a post's title
- can see some of the post's body
- can see the first comments on a post
- can see how many comments a post has
- can see how many likes a post has
- can see a section for pagination if there are more posts than fit on the view

### post/show
can see the post's title
can see how many comments it has
can see how many likes it has
can see the post body
can see the username of each commentor
can see the comment each commentor left

**Comment by RileyManda** [comment]
The following iterations have been implemented as requested by the reviewer

### Iterations


## post/index

- can see the user's profile picture
- can see the user's username
- can see the number of posts the user has written
- can see a post's title
- can see some of the post's body
- can see the first comments on a post
- can see how many comments a post has
- can see how many likes a post has
- can see a section for pagination if there are more posts than fit on the view

## post/show

- can see the post's title
- can see how many comments it has
- can see how many likes it has
- can see the post body
- can see the username of each commentor
- can see the comment each commentor left

**Comment by RileyManda** [comment]
xcx

**Review by AbelG101 — APPROVED** [review]
# Hi @RileyManda& @rloterh 👋, 

**Your project is complete!  There is nothing else to say other than... it's time to merge it. 
Congratulations! 🎉 🥳🥇**

## Highlights
- All Tests are properly added ✔
- Correct use of Git Flow ✔
- All linter checks are passing ✔
- Descriptive PR title with a great summary ✔

## Optional suggestions
- N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**